### PR TITLE
build: untangle `UNITTESTS` and `DEBUGBUILD` macros

### PR DESCRIPTION
**Pull request by vszakats**
- fix `DEBUGBUILD` guards that should be `UNITTESTS`, in libcurl
  code used by unit tests.
- fix guards for libcurl functions used in unit tests only.
- sync `UNITTEST` attribute between declarations and definitions.
- drop `DEBUGBUILD` guard from test `unit2600`.
- fix guards for libcurl HSTS code used by both a unit test (`unit1660`)
  and `test0446`.
- update an existing AppVeyor CI job to test the issues fixed.

This fixes building tests with `CURLDEBUG` enabled but `DEBUGBUILD`
disabled. This can happen when building tests with CMake with
`ENABLE_DEBUG=ON` in Release config, or with `ENABLE_CURLDEBUG=ON`
and _without_ `ENABLE_DEBUG=ON`. Possibly also with autotools
when using `--enable-curldebug` without `--enable-debug`.

Test results:
- before: https://ci.appveyor.com/project/curlorg/curl/builds/49835609
- after: https://ci.appveyor.com/project/curlorg/curl/builds/49839255
  (the two failures are unrelated, subject to PR #13705)

Ref: #13592 (issue discovery)
Ref: #13689 (CI testing this PR with `DEBUGBUILD`/`CURLDEBUG` combinations)
Closes #13694

### CODE REVIEW

File: lib/hsts.c [hunk 1]
```diff
@@ -54,7 +54,7 @@
 #define MAX_HSTS_DATELENSTR "64"
 #define UNLIMITED "unlimited"
 
-#ifdef DEBUGBUILD
+#if defined(DEBUGBUILD) || defined(UNITTESTS)
```
Comment (bagder) [inline lib/hsts.c:57]: Why is this needed?

The unit tests assume that `DEBUGBUILD` is defined as a prereq for running. This change seems to say that you can build with `UNITTESTS` without `DEBUGBUILD` ?

**Comment by bagder** [comment]
> This fixes building tests with CURLDEBUG enabled but DEBUGBUILD disabled

I wish we could drop the distinction completely and just use one of them. I can't remember the difference, I don't think anyone uses them separately. Am I wrong?

File: lib/hsts.c (hunk 1 above)
Comment (vszakats) [reply lib/hsts.c:57]: Unit tests (and any other test) are _building_ fine without either `DEBUGBUILD` or `CURLDEBUG`. (That is after fixing the places where `DEBUGBUILD` was used to enable functions for unit tests.)

This is an outlier case where the `deltatime` variable of a debug feature (`CURL_TIME`) is actually required for building a unit test.

**Comment by vszakats** [comment]
> > This fixes building tests with CURLDEBUG enabled but DEBUGBUILD disabled
> 
> I wish we could drop the distinction completely and just use one of them. I can't remember the difference, I don't think anyone uses them separately. Am I wrong?

They are not the same thing, but their names don't make it trivial to figure out which does what exactly:
- `DEBUGBUILD`: curl debug features, such as extra output, and envs that allow controlling internals.
- `CURLDEBUG`: memory tracking feature, aka 'TrackMemory'. It's an intrusive option that redefines a lot of internal functions, somewhat fragile.

In CMake, `DEBUGBUILD` always enabled `CURLDEBUG`, but the latter could also be enabled separately. Autotools looks somewhat similar. Both offer separate options to enable them. `Makefile.mk` also has independent switches. winbuild enables `DEBUGBUILD`, but never `CURLDEBUG`. It's probably rare indeed that someone is concsiously using them separately though.

Yet, they are separate and possible to use separately, with quite a bit of breakage around the edges.

The goal of this patch isn't really to untangle these two from each other, but to fix to use the `UNITTESTS` macro to enable internal functions and features required for building _unit tests_. This is mandatory for a successful build, unlike `DEBUGBUILD` and `CURLDEBUG` which are optional for building tests. (`DEBUGBUILD` seems required to _run_ them, but I'd have to test this.)

Merging `DEBUGBUILD` and `CURLDEBUG` is an option, but something for separate PR(s).

This PR is an attempt t fix all existing build combinations.

**Comment by vszakats** [comment]
There seem to be a few suspicious cases, where `DEBUGBUILD` and `CURLDEBUG` are mixed up in the code, too.
E.g. `src/tool_operate.c` guards some debug features with `CURLDEBUG`. There is also one case in `src/tool_getparam.c` + `src/tool_cfgable.h`. On a quick glance, in `lib` almost all `CURLDEBUG` should be `DEBUGBUILD`.

I think we might consider renaming `CURLDEBUG` to `CURL_TRACK_MEMORY` or something to make its purpose clear and avoid future mis-uses.

With the list of related open PRs in the air, I'd prefer merging these first, then continue tackling more. Two PRs are blocked by this one: #13705 and #13698.

**Comment by jay** [comment]
> I wish we could drop the distinction completely and just use one of them. I can't remember the difference, I don't think anyone uses them separately. Am I wrong?

As Viktor mentions above, yes they are separate because CURLDEBUG is (supposed to be) the guard for the trackmemory feature. Typically I run debug builds of curl with DEBUGBUILD but not CURLDEBUG.

> I think we might consider renaming CURLDEBUG to CURL_TRACK_MEMORY / CURL_DEBUG_MEMORY or something to make its purpose clear and avoid future mis-uses. For DEBUGBUILD, CURL_DEBUG (or anything starting with CURL_ might also work better.

I'm with you on the first part (maybe something shorter?) but not the second part. Renaming DEBUGBUILD to CURL_DEBUG is confusing to me since we've been using CURLDEBUG for memory tracking.

**Comment by vszakats** [comment]
> I'm with you on the first part (maybe something shorter?) but not the second part. Renaming DEBUGBUILD to CURL_DEBUG is confusing to me since we've been using CURLDEBUG for memory tracking.

Indeed, I agree.

_If_ we want to rename them (or one of them), we can brainstorm for new names that are short, tell the purpose, and without confusion. Starting with `CURL_` would be a plus for namespacing I think. (`DEBUGBUILD` might be something used by a dependency, though if it didn't happen for 2 decades, it's not likely to happen now.)

Anyway, after tidying this up, I think it will be a little bit harder to make a mistake by copy-pasting existing code or looking up existing uses. `CURLDEBUG` is used in just a few specific places after these updates.

As for `DEBUGBUILD` and `UNITTESTS`, they remain easy to confuse. The former isn't required to build tests, the latter is.
It makes the purpose of these debug features easier to grasp, but it's debatable if this is a strong enough purpose, esp. if new code continues to get them mixed up. I'm 100% yet, but merging `UNITTESTS` into `DEBUGBUILD` _might_ allow dropping the `curlu` library, by including all these in the debug libcurl lib, which could then be used by unit tests.

**Comment by jay** [comment]
I can see UNITTESTS being dependent on DEBUGBUILD but not the other way around
